activestorage/lib/active_storage/service/azure_storage_service.rb

```diff
@@ -34,16 +34,26 @@ def download(key, &block)
         end
       else
         instrument :download, key: key do
-          _, io = blobs.get_blob(container, key)
-          io.force_encoding(Encoding::BINARY)
+          begin
+            _, io = blobs.get_blob(container, key)
+            io.force_encoding(Encoding::BINARY)
+          rescue Azure::Core::Http::HTTPError => e
+            raise e unless e.type == "BlobNotFound"
+            raise ActiveStorage::FileNotFoundError
+          end
         end
       end
     end
 
     def download_chunk(key, range)
       instrument :download_chunk, key: key, range: range do
-        _, io = blobs.get_blob(container, key, start_range: range.begin, end_range: range.exclude_end? ? range.end - 1 : range.end)
-        io.force_encoding(Encoding::BINARY)
+        begin
+          _, io = blobs.get_blob(container, key, start_range: range.begin, end_range: range.exclude_end? ? range.end - 1 : range.end)
+          io.force_encoding(Encoding::BINARY)
+        rescue Azure::Core::Http::HTTPError => e
+          raise e unless e.type == "BlobNotFound"
+          raise ActiveStorage::FileNotFoundError
+        end
```

Comment:
Can we extract a private `handle_errors` method to avoid duplicating the error handling logic in `download` and `download_chunk`?

```ruby
def download_chunk(key, range)
  instrument :download_chunk, key: key, range: range do
    handle_errors do
      _, io = blobs.get_blob(container, key, start_range: range.begin, end_range: range.exclude_end? ? range.end - 1 : range.end)
      io.force_encoding(Encoding::BINARY)
    end
  end
end
```

Reply:
I was half tempted to put such a private method on the parent class (taking a service specific exception as an argument) but worried it was gonna be too much magic. Is this just needed because of the extra type checking needed for Azure errors? (Related bug fix PR #33667 that should probably do it the same way…)

Reply:
I wouldn’t put it on the parent class. It’s a service-specific method.